src/extension/popup/PopupCommon.re

```diff
@@ -31,24 +31,64 @@ let makeContentHash text => "#" ^ Util.btoa text;
 let generateShareableLink text => "https://reasonml.github.io/reason-tools/popup.html" ^ text;
 
 let getInputFromUrl () => {
-  let text = Location.hash |> Js.String.sliceToEnd from::1 |> Util.atob;
+  let text = LocalDom.Location.hash |> Js.String.sliceToEnd from::1 |> Util.atob;
   if (Str.isEmpty text) {
     Promise.reject ()
   } else {
     Promise.resolve text
   }
 };
 
-let init ::getSelection=rejectedPromise ::getLatestInput=rejectedPromise ::onOpen ::refmt () => {
-  let rec inputChanged input => {
+let init
+    ::getSelection=rejectedPromise
+    ::getLatestInput=rejectedPromise
+    ::onOpen
+    refmt::(
+      refmt:
+        Js.String.t =>
+        inLang::string? =>
+        inType::string? =>
+        outLang::string? =>
+        (string => option string => option string => 'a) =>
+        unit
+    )
+    () => {
+  let rec inLangChanged input e => {
+    let hash = makeContentHash input;
+    let link = generateShareableLink hash;
+    let inLang = e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value;
+    refmt input ::inLang (fun outText inLang outLang => render input outText inLang outLang link)
+  }
+  and outLangChanged input e => {
+    let hash = makeContentHash input;
+    let link = generateShareableLink hash;
+    let outLang = e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value;
+    refmt input ::outLang (fun outText inLang outLang => render input outText inLang outLang link)
```

Comment:
There's not a whole lot of difference between `inLangChanged` and `outLangChanged`. A couple of helpers, one that wraps the refmt call and computes `link` and `hash`, and one that retrieves the value would be good perhaps.